File: adapt_cc/src/adapt/dom/cfi.cc
#include "adapt/dom/cfi.h"

#include <stdlib.h>

#include "adapt/base/rc_map.h"
#include "adapt/base/rc_list.h"

namespace adapt {
namespace dom {

namespace {

void escape(Writer* writer, const rc_string& text) {
  const char* s = text->utf8();
  while (true) {
    char c = *(s++);
    switch (c) {
      case '\0' :
        break;
      case '[' :
      case ']' :
      case '(' :
      case ')' :
      case ',' :
      case '=' :
      case ';' :
      case '^' :
        *writer << '^' << c;
        break;
      default:
        *writer << c;
    }
  }
}

rc_list<rc_string> parse_value(const char** sp) {
  StringWriter w;
  std::vector<rc_string> values;
  const char* s = *sp;
  while (*s != ';' && *s != ']') {
    if (*s == ',') {
      values.push_back(w.to_string());
      w.clear();
    }
    if (*s == '^') {
      ++s;
      if (!*s) {
        return rc_list<rc_string>::empty();
      }
    }
    w << *s;
    ++s;
  }
  values.push_back(w.to_string());
  *sp = s;
  return rc_list<rc_string>(values);
}

rc_map<rc_list<rc_string>> parse_ext(const char** sp) {
  const char* s = *sp;
  std::map<rc_qname, rc_list<rc_string>> ext;
  while (*s == ';') {
    ++s;
    const char *p = s;
    while (*s != '=' && *s != ';' && *s != ']') {
      ++s;
    }
    if (*s != '=') {
      break;
    }
    rc_qname name(p, s - p);
    ++s;
    rc_list<rc_string> list = parse_value(&s);
    if (list->empty()) {
      break;
    }
    ext[name] = list;
  }
  *sp = s;
  return rc_map<rc_list<rc_string>>(ext);
}

SideBias get_bias(const rc_map<rc_list<rc_string>>& ext) {
  rc_list<rc_string> bias_str = ext[rc_qname::atom(ID_s)];
  if (!bias_str.is_null() && bias_str->size() == 1) {
    if (bias_str[0] == "b") {
      return BIAS_BEFORE;
    } else if (bias_str[0] == "a"){
      return BIAS_AFTER;
    }
  }
  return BIAS_NONE;
}

class RefStep : public CFIStep {
 public:
  virtual bool apply_to(Position* pos, rc_string* err) const {
    return false;
  }

  virtual void write_to(Writer* writer) const override {
    *writer << '!';
  }
};

class ChildStep : public CFIStep {
 public:
  ChildStep(size_t index, const rc_qname& id, SideBias side_bias)
      : index_(index), id_(id), side_bias_(side_bias) {}

  virtual bool apply_to(Position* pos, rc_string* err) const {
    if (index_ == 1) {
      // text content
      pos->offset = 1;
      pos->after = false;
      return true;
    }
    size_t element_index = index_ / 2 - 1;
    const dom::Children* children = pos->element->children();
    if (!children || element_index >= children->size()) {
      *err = "E_CFI_NO_CHILD";
      return true;
    }
    pos->element = children->element(element_index);
    pos->offset = 0;
    pos->after = (index_ & 1) != 0;
    if (!id_.empty()) {
      rc_map<rc_value> attrs = pos->element->attributes();
      if (attrs[rc_qname::atom(ID_id)] != id_ &&
          attrs[rc_qname::atom(ID_XML_id)] != id_) {
        *err = "E_CFI_ID_MISMATCH";
      }
    }
    pos->side_bias = side_bias_;
    return true;
  }

  virtual void write_to(Writer* writer) const override {
    *writer << '/' << index_;
    if (!id_.empty() || side_bias_ != BIAS_NONE) {
      *writer << '[';
      if (!id_.empty()) {
        *writer << (id_->to_string());
      }
      if (side_bias_ != BIAS_NONE) {
        *writer << (side_bias_ == BIAS_AFTER ? ";s=a" : ";s=b");
      }
      *writer << ']';
    }
  }

 private:
  size_t index_;
  rc_qname id_;
  SideBias side_bias_;
};

class OffsetStep : public CFIStep {
 public:
  OffsetStep(dom::node_offset_t offset, const rc_string& text_before,
      const rc_string& text_after, SideBias side_bias)
          : offset_(offset), text_before_(text_before), text_after_(text_after),
            side_bias_(side_bias) {}

  virtual bool apply_to(Position* pos, rc_string* err) const {
    pos->offset += offset_;
    pos->side_bias = side_bias_;
    // TODO: text before/after validation
    return true;
  }

  virtual void write_to(Writer* writer) const override {
    *writer << ':' << offset_;
    if (!text_before_->empty() || !text_after_->empty()
        || side_bias_ != BIAS_NONE) {
      *writer << '[';
      if (!text_before_->empty() || !text_after_->empty()) {
        if (!text_before_->empty()) {
          escape(writer, text_before_);
        }
        *writer << ',';
        if (!text_after_->empty()) {
          escape(writer, text_after_);
        }
      }
      if (side_bias_ != BIAS_NONE) {
        *writer << (side_bias_ == BIAS_AFTER ? ";s=a" : ";s=b");
      }
      *writer << ']';
    }
  }

  dom::node_offset_t offset_;
  rc_string text_before_;
  rc_string text_after_;
  SideBias side_bias_;
};

}

rc_ref<CFIFragment> CFIFragment::from(const rc_string& str) {
  const char* s = str->utf8();
  if (*s == '#') {
    ++s;
  }
  if (strncmp(s, "epubcfi(", 8) != 0) {
    fprintf(stderr, "E_CFI_NOT_CFI\n");
    return rc_ref<CFIFragment>();
  }
  s += 8;
  std::vector<rc_ref<CFIStep>> steps;
  while(true) {
    switch (*s) {
      case ')':
        if (s[1] != '\0') {
          fprintf(stderr, "E_CFI_JUNK_AFTER_END\n");
          return rc_ref<CFIFragment>();
        }
        return new CFIFragment(steps);
      case '/': {
        ++s;
        if ('0' > *s || *s > '9') {
          fprintf(stderr, "E_CFI_NUMBER_EXPECTED\n");
          return rc_ref<CFIFragment>();
        }
        char* end;
        long index = ::strtol(s, &end, 10);
        if (index <= 1) {
          fprintf(stderr, "E_CFI_POSITIVE_NUMBER_EXPECTED\n");
          return rc_ref<CFIFragment>();
        }
        rc_qname id;
        rc_map<rc_list<rc_string>> ext = rc_map<rc_list<rc_string>>::empty();
        s = end;
        if (*s == '[') {
          ++s;
          const char *p = s;
          while (*s && *s != ']' && *s != ';') {
            ++s;
          }
          if (p > s) {
            id = rc_qname(p, s - p);
          }
          if (*s == '\0') {
            fprintf(stderr, "E_CFI_INCOMPLETE\n");
            return rc_ref<CFIFragment>();
          }
          if (*s == ';') {
            ext = parse_ext(&s);
            if (*s != ']') {
              fprintf(stderr, "E_CFI_SYNTAX\n");
              return rc_ref<CFIFragment>();
            }
            ++s;
          }
        }
        steps.push_back(new ChildStep(index, id, get_bias(ext)));
        break;
      }
      case ':': {
        ++s;
        if ('0' > *s || *s > '9') {
          fprintf(stderr, "E_CFI_NUMBER_EXPECTED\n");
          return rc_ref<CFIFragment>();
        }
        char* end;
        long offset = ::strtol(s, &end, 10);
        s = end;
        rc_list<rc_string> text = rc_list<rc_string>::empty();
        rc_map<rc_list<rc_string>> ext = rc_map<rc_list<rc_string>>::empty();
        if (*s == '[') {
          ++s;
          text = parse_value(&s);
          if (*s == ';') {
            ext = parse_ext(&s);
          }
          if (*s != ']') {
            fprintf(stderr, "E_CFI_SYNTAX\n");
            return rc_ref<CFIFragment>();
          }
          ++s;
        }
        rc_string text_before = text->size() >= 1 ? text[0] : rc_string();
        rc_string text_after = text->size() >= 2 ? text[1] : rc_string();
        steps.push_back(new OffsetStep(offset, text_before,
            text_after, get_bias(ext)));
        break;
      }
      case '!':
        s++;
        steps.push_back(new RefStep());
        break;
      case '~':
      case '@':
        // Time/space terminus; only useful for highlights/selections which are not yet
        // supported, skip for now.
        s += strlen(s) - 1;
        break;
      default:
        fprintf(stderr, "E_CFI_UNKNOWN_STEP\n");
        return rc_ref<CFIFragment>();
    }
  }
}

rc_ref<CFIFragment> CFIFragment::from(const Position& pos) {
  std::vector<rc_ref<CFIStep>> steps;
  SideBias side_bias = pos.side_bias;
  bool next = false;
  if (pos.after || pos.offset > 0) {
    assert(pos.ref.is_null());
    // Text node
    size_t off = pos.offset;
    rc_string text;
    if (pos.after) {
      next = true;
      text = pos.element->following_text();
    } else {
      --off;
      text = pos.element->leading_text();
    }
    const size_t sz = 8;
    size_t len = text->utf16_length();
    rc_string text_before = text->utf16_substr(std::max(off, sz) - sz, off);
    rc_string text_after = text->utf16_substr(off, std::min(len, off + sz));
    steps.push_back(new OffsetStep(off, text_before, text_after, side_bias));
    if (!pos.after) {
      steps.push_back(new ChildStep(0, rc_qname(), BIAS_NONE));
    }
    side_bias = BIAS_NONE;
  }
  dom::DOM::iterator element = pos.element;
  while(element.has_parent()) {
    size_t index = 2 * (element.child_index() + 1);
    rc_qname id;
    if (next) {
      ++index;
    } else {
      const rc_map<rc_value>& attr = element->attributes();
      rc_value idv = attr[rc_qname::atom(ID_id)];
      if (idv.is_null()) {
        idv = attr[rc_qname::atom(ID_XML_id)];
      }
      if (!idv.is_null()) {
        id = rc_cast<rc_qname>(idv);
      }
    }
    steps.push_back(new ChildStep(index, id, side_bias));
    side_bias = BIAS_NONE;
    element = element.parent();
  }
  std::reverse(steps.begin(), steps.end());
  if (!pos.ref.is_null() && pos.ref->is<rc_ref<CFIFragment>>()) {
    CFIFragment* ref = static_cast<CFIFragment*>(pos.ref.raw_ptr());
    steps.push_back(new RefStep());
    for (const rc_ref<CFIStep>& step : ref->steps_) {
      steps.push_back(step);
    }
  }
  return new CFIFragment(steps);
}

bool CFIFragment::navigate(dom::DOM::iterator root, Position* pos) {
  rc_string err;
  pos->element = root;
  pos->after = false;
  pos->offset = 0;
  pos->side_bias = BIAS_NONE;
  pos->ref = rc_value();
  for (auto it = steps_.begin(); it != steps_.end(); ++it) {
    const rc_ref<CFIStep>& step = *it;
    if (!step->apply_to(pos, &err)) {
      std::vector<rc_ref<CFIStep>> steps;
      for (++it; it != steps_.end(); ++it) {
        steps.push_back(*it);
      }
      pos->ref = new CFIFragment(steps);
      break;
    }
    if (!err->empty()) {
      return false;
    }
  }
  return true;
}

void CFIFragment::write_to(Writer* writer) const {
  *writer << "#epubcfi(";
  for (const rc_ref<CFIStep> step : steps_) {
    step->write_to(writer);
  }
  *writer << ')';
}

CFIFragment::CFIFragment(const std::vector<rc_ref<CFIStep>>& steps)
    : steps_(steps) {}

}
}

File: js/adapt/cfi.js
/**
 * Copyright 2013 Google, Inc.
 * @fileoverview Support for EPUB Canonical Fragment Identifiers.
 */
goog.provide('adapt.cfi');

goog.require('adapt.base');

/**
 * @typedef {{
 * 		node: Node,
 * 		offset: number,
 * 		after: boolean,
 *		sideBias: ?string,
 *		ref: adapt.cfi.Fragment
 * }}
 */
adapt.cfi.Position;

/**
 * @param {Node} node
 * @return {?string}
 */
adapt.cfi.getId = function(node) {
	if (node.nodeType == 1) {
		var idtxt = (/** @type {Element} */ (node)).getAttribute("id");
		if (idtxt && idtxt.match(/^[-a-zA-Z_0-9.\u007F-\uFFFF]+$/)) {
			return idtxt;
		}
	}
	return null;
};

/**
 * @param {string} ch
 * @return {string}
 */
adapt.cfi.escapeChar = function(ch) {
	return "^" + ch;
};

/**
 * @param {string} str
 * @return {string}
 */
adapt.cfi.escape = function(str) {
	return str.replace(/[\[\]\(\),=;^]/g, adapt.cfi.escapeChar);
};

/**
 * @param {string} str
 * @return {string}
 */
adapt.cfi.unescapeChar = function(str) {
	return str.substr(1);
};


/**
 * @param {string} str
 * @return {string}
 */
adapt.cfi.unescape = function(str) {
	if (!str)
		return str;
	return str.replace(/\^[\[\]\(\),=;^]/g, adapt.cfi.unescapeChar);
};

/**
 * @param {string} extstr
 * @return {string|Array.<string>}
 */
adapt.cfi.parseExtVal = function(extstr) {
	var result = [];
	do {
		var r = extstr.match(/^(\^,|[^,])*/);
		var p = adapt.cfi.unescape(r[0]);
		extstr = extstr.substr(r[0].length + 1);
		if (!extstr && !result.length)
			return p;
		result.push(p);
	} while (extstr);
	return result;
};

/**
 * @param {string} extstr
 * @return {Object.<string,string|Array.<string>>}
 */
adapt.cfi.parseExt = function(extstr) {
	var ext = {};
	while(extstr) {
		var r = extstr.match(/^;([^;=]+)=(([^;]|\^;)*)/);
		if (!r) {
			return ext;
		}
		ext[r[1]] = adapt.cfi.parseExtVal(r[2]);
		extstr = extstr.substr(r[0].length);
	}
	return ext;
};

/**
 * @interface
 */
adapt.cfi.Step = function() {};

/**
 * @param {adapt.base.StringBuffer} sb
 * @return {void}
 */
adapt.cfi.Step.prototype.appendTo = function(sb) {};

/**
 * @param {adapt.cfi.Position} pos
 * @return {boolean}
 */
adapt.cfi.Step.prototype.applyTo = function(pos) {};

/**
 * @constructor
 * @implements {adapt.cfi.Step}
 */
adapt.cfi.RefStep = function() {};

/**
 * @param {adapt.base.StringBuffer} sb
 */
adapt.cfi.RefStep.prototype.appendTo = function(sb) {
	sb.append("!");
};

/**
 * @override
 */
adapt.cfi.RefStep.prototype.applyTo = function(pos) {
	return false;
};

/**
 * @constructor
 * @param {number} index
 * @param {?string} id
 * @param {?string} sideBias
 * @implements {adapt.cfi.Step}
 */
adapt.cfi.ChildStep = function(index, id, sideBias) {
	/** @const */ this.index = index;
	/** @const */ this.id = id;
	/** @const */ this.sideBias = sideBias;
};

/**
 * @override
 */
adapt.cfi.ChildStep.prototype.appendTo = function(sb) {
	sb.append("/");
	sb.append(this.index.toString());
	if (this.id || this.sideBias) {
		sb.append("[");
		if (this.id) {
			sb.append(this.id);
		}
		if (this.sideBias) {
			sb.append(";s=");
			sb.append(this.sideBias);
		}
		sb.append("]");
	}
};

/**
 * @override
 */
adapt.cfi.ChildStep.prototype.applyTo = function(pos) {
	if (pos.node.nodeType != 1) {
		throw new Error("E_CFI_NOT_ELEMENT");
	}
	var elem = /** @type {Element} */ (pos.node);
	var childElements = elem.children;
	var childElementCount = childElements.length;
	var child;
	var childIndex = Math.floor(this.index / 2) - 1;
	if (childIndex < 0 || childElementCount == 0) {
		child = elem.firstChild;
		pos.node = child || elem;
	} else {
		child = childElements[Math.min(childIndex, childElementCount - 1)];
		if (this.index & 1) {
			var next = child.nextSibling;
			if (!next || next.nodeType == 1) {
				pos.after = true;
			} else {
				child = next;
			}
		}
		pos.node = child;
	}
	if (this.id && (pos.after || this.id != adapt.cfi.getId(pos.node))) {
		throw new Error("E_CFI_ID_MISMATCH");		
	}
	pos.sideBias = this.sideBias;
	return true;
};


/**
 * @constructor
 * @param {number} offset
 * @param {?string} textBefore
 * @param {?string} textAfter
 * @param {?string} sideBias
 * @implements {adapt.cfi.Step}
 */
adapt.cfi.OffsetStep = function(offset, textBefore, textAfter, sideBias) {
	/** @const */ this.offset = offset;
	/** @const */ this.textBefore = textBefore;
	/** @const */ this.textAfter = textAfter;
	/** @const */ this.sideBias = sideBias;
};

adapt.cfi.OffsetStep.prototype.applyTo = function(pos) {
	if (this.offset > 0 && !pos.after) {
		var offset = this.offset;
		var node = pos.node;
		while (true) {
			var nodeType = node.nodeType;
			if (nodeType == 1) {
				break;
			}
			var next = node.nextSibling;
			if (3 <= nodeType && nodeType <= 5) {
				var textLength = node.textContent.length;
				if (offset <= textLength) {
					break;
				}
				if (!next) {
					offset = textLength;
					break;
				}
				offset -= textLength;
			}
			if (!next) {
				offset = 0;
				break;
			}
			node = next;
		}
		pos.node = node;
		pos.offset = offset;
	}
	pos.sideBias = this.sideBias;
	return true;	
};

/**
 * @override
 */
adapt.cfi.OffsetStep.prototype.appendTo = function(sb) {
	sb.append(":");
	sb.append(this.offset.toString());
	if (this.textBefore || this.textAfter || this.sideBias) {
		sb.append("[");
		if (this.textBefore || this.textAfter) {
			if (this.textBefore) {
				sb.append(adapt.cfi.escape(this.textBefore));
			}
			sb.append(",");
			if (this.textAfter) {
				sb.append(adapt.cfi.escape(this.textAfter));
			}
		}
		if (this.sideBias) {
			sb.append(";s=");
			sb.append(this.sideBias);
		}
		sb.append("]");
	}
};


/**
 * @constructor
 */
adapt.cfi.Fragment = function() {
	/** @type {Array.<adapt.cfi.Step>} */ this.steps = null;
};

/**
 * @param {string} fragstr
 * @return {void}
 */
adapt.cfi.Fragment.prototype.fromString = function(fragstr) {
	var r = fragstr.match(/^#?epubcfi\((.*)\)$/);
	if (!r) {
		throw new Error("E_CFI_NOT_CFI");
	}
	var str = r[1];
	var i = 0;
	var steps = [];
	while(true) {
		switch (str.charAt(i)) {
		case "/":
			i++;
			r = str.substr(i).match(/^(0|[1-9][0-9]*)(\[([-a-zA-Z_0-9.\u007F-\uFFFF]+)(;([^\]]|\^\])*)?\])?/);
			if (!r) {
				throw new Error("E_CFI_NUMBER_EXPECTED");				
			}
			i += r[0].length;
			var index = parseInt(r[1], 10);
			var id = r[3];
			var ext = adapt.cfi.parseExt(r[4]);
			steps.push(new adapt.cfi.ChildStep(index, id, adapt.base.asString(ext["s"])));
			break;
		case ":":
			i++;
			r = str.substr(i).match(/^(0|[1-9][0-9]*)(\[((([^\];,]|\^[\];,])*)(,(([^\];,]|\^[\];,])*))?)(;([^]]|\^\])*)?\])?/);
			if (!r) {
				throw new Error("E_CFI_NUMBER_EXPECTED");				
			}
			i += r[0].length;
			var offset = parseInt(r[1], 10);
			var textBefore = r[4];
			if (textBefore) {
				textBefore = adapt.cfi.unescape(textBefore);
			}
			var textAfter = r[7];
			if (textAfter) {
				textAfter = adapt.cfi.unescape(textAfter);
			}
			var ext = adapt.cfi.parseExt(r[10]);
			steps.push(new adapt.cfi.OffsetStep(offset, textBefore, textAfter, adapt.base.asString(ext["s"])));
			break;			
		case "!":
			i++;
			steps.push(new adapt.cfi.RefStep());
			break;
		case "~":
		case "@":
			// Time/space terminus; only useful for highlights/selections which are not yet
			// supported, skip for now.
			// fall through
		case "":
			this.steps = steps;
			return;
		default:
			throw new Error("E_CFI_PARSE_ERROR");				
		}
	}
};

/**
 * @param {Document} doc
 * @return {adapt.cfi.Position}
 */
adapt.cfi.Fragment.prototype.navigate = function(doc) {
	var pos = {node:doc.documentElement, offset:0, after:false, sideBias:null, ref:null};
	for (var i = 0; i < this.steps.length; i++) {
		if (!this.steps[i].applyTo(pos)) {
			if (++i < this.steps.length) {
				pos.ref = new adapt.cfi.Fragment();
				pos.ref.steps = this.steps.slice(i);
			}
			break;
		}
	}
	return pos;
};

/**
 * @param {string} text
 * @param {boolean} after
 * @return {string}
 */
adapt.cfi.Fragment.prototype.trim = function(text, after) {
	return text.replace(/\s+/g, " ").match(
			after ? /^[ -\uD7FF\uE000-\uFFFF]{0,8}/ : /[ -\uD7FF\uE000-\uFFFF]{0,8}$/
		)[0].replace(/^\s/, "").replace(/\s$/, "");
};

/**
 * Initialize from a node and an offset.
 * @param {Node} node
 * @param {number} offset
 * @param {boolean} after
 * @param {?string} sideBias
 */
adapt.cfi.Fragment.prototype.prependPathFromNode = function(node, offset, after, sideBias) {
	var steps = [];
	var parent = node.parentNode;
	var textBefore = "";
	var textAfter = "";
	while(node) {
		switch (node.nodeType) {
		case 3:  // Text nodes
		case 4:
		case 5:
			var text = node.textContent;
			var textLength = text.length;
			if (after) {
				offset += textLength;
				if (!textBefore) {
					textBefore = text;
				}
			} else {
				if (offset > textLength) {
					offset = textLength;
				}
				after = true;
				textBefore = text.substr(0, offset);
				textAfter = text.substr(offset);
			}
			node = node.previousSibling;
			continue;
		case 8:  // Comment Node
            node = node.previousSibling;
			continue;
		}
		break;
	}
	if (offset > 0 || textBefore || textAfter) {
		textBefore = this.trim(textBefore, false);
		textAfter = this.trim(textAfter, true);
 		steps.push(new adapt.cfi.OffsetStep(offset, textBefore, textAfter, sideBias));
 		sideBias = null;
	}
	while (parent) {
		if (!parent || parent.nodeType == 9) {
			break;
		}
 		var id = after ? null : adapt.cfi.getId(node);
 		var index = after ? 1 : 0;
 		while (node) {
 			if (node.nodeType == 1) {
 				index += 2;
 			}
 			node = node.previousSibling;
 		}
 		steps.push(new adapt.cfi.ChildStep(index, id, sideBias));
 		sideBias = null;
		node = parent;
		parent = parent.parentNode;
		after = false;
	}
	steps.reverse();
	if (this.steps) {
		steps.push(new adapt.cfi.RefStep());
		this.steps = steps.concat(this.steps);
	} else {
		this.steps = steps;
	}
};

adapt.cfi.Fragment.prototype.toString = function() {
	if (!this.steps)
		return "";
	var sb = new adapt.base.StringBuffer();
	sb.append("epubcfi(");
	for (var i = 0; i < this.steps.length; i++) {
		this.steps[i].appendTo(sb);
	}
	sb.append(")");
	return sb.toString();	
};